src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -874,7 +878,7 @@ private void ReportNullabilityMismatchInAssignment(SyntaxNode syntaxNode, object
                     // https://github.com/dotnet/roslyn/issues/31395: We should copy all tracked state from `value` regardless of
                     // BoundNode type but we'll need to handle cycles (see NullableReferenceTypesTests.Members_FieldCycle_07).
                     // For now, we copy a limited set of BoundNode types that shouldn't contain cycles.
-                    if ((targetType.IsReferenceType && (isSupportedReferenceTypeValue(value) || targetType.TypeSymbol.IsAnonymousType)) ||
+                    if (((targetType.IsReferenceType || targetType.TypeKind == TypeKind.TypeParameter) && (isSupportedReferenceTypeValue(value) || targetType.TypeSymbol.IsAnonymousType)) ||
```

Comment:
>targetType.TypeKind == TypeKind.TypeParameter [](start = 56, length = 45)

I found this change surprising. Why isn't `targetType.IsReferenceType` sufficient?

Reply:
Unconstrained type parameters, and type parameters constrained to value types, were not handled here nor in the `else` previously.

---
In reply to: [265721470](https://github.com/dotnet/roslyn/pull/34087#discussion_r265721470) [](ancestors = 265721470)